// @nebular/firebase-auth/esm2015/strategies/base/firebase-identity-provider-strategy.options.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import { NbAuthJWTToken, NbAuthStrategyOptions, } from '@nebular/auth';
export class NbFirebaseIdentityProviderStrategyOptions extends NbAuthStrategyOptions {
    constructor() {
        super(...arguments);
        this.token = {
            class: NbAuthJWTToken,
        };
        this.logout = {
            redirect: {
                success: '/',
                failure: null,
            },
            defaultErrors: ['Something went wrong, please try again.'],
            defaultMessages: ['You have been successfully logged out.'],
        };
        this.authenticate = {
            redirect: {
                success: '/',
                failure: null,
            },
            defaultErrors: ['Something went wrong, please try again.'],
            defaultMessages: ['You have been successfully authenticated.'],
        };
        this.errors = {
            key: 'message',
            getter: (module, res, options) => options[module].defaultErrors,
        };
        this.messages = {
            key: 'message',
            getter: (module, res, options) => {
                return options[module].defaultMessages;
            },
        };
        this.scopes = [];
        this.customParameters = {};
    }
}
;
